src/peft/utils/loftq_utils.py

```diff
@@ -227,3 +233,139 @@ def loftq_init(weight: Union[torch.Tensor, torch.nn.Parameter], num_bits: int, r
     lora_A, lora_B = R, L
 
     return dequantized_weight.to(device=device, dtype=dtype), lora_A, lora_B
+
+
+def _check_model_path_loftq(model_path, peft_model):
+    """Check if there is a local model file corresponding to the given model, then format it properly."""
+    if model_path is None:
+        try:
+            model_path = snapshot_download(peft_model.base_model.config._name_or_path, local_files_only=True)
+        except AttributeError as exc:
+            raise ValueError(
+                "The provided model does not appear to be a transformers model. In this case, you must pass the "
+                "model_path to the safetensors file."
+            ) from exc
+        except LocalEntryNotFoundError as exc:
+            raise ValueError("The model.safetensors file must be present on disk, but it could not be found.") from exc
+
+    suffix = "model.safetensors"
+    if not model_path.endswith(suffix):
+        model_path = os.path.join(model_path, suffix)
+    if not os.path.exists(model_path):
+        raise FileNotFoundError(
+            f"Could not find file {model_path}, ensure that there is a safetensors file of the model."
+        )
+
+    return model_path
+
+
+@torch.no_grad()
+def _loftq_init_new(qweight, weight, num_bits: int, reduced_rank: int):
+    if num_bits != 4:
+        raise ValueError("Only 4 bit quantization supported at the moment.")
+    if not is_bnb_4bit_available():
+        raise ValueError("bitsandbytes 4bit quantization is not available.")
+
+    compute_device = "cuda"
+    dequantized_weight = bnb.functional.dequantize_4bit(qweight.data, qweight.quant_state)
+
+    weight = weight.to(device=compute_device, dtype=torch.float32)
+    residual = weight - dequantized_weight
+    torch.cuda.empty_cache()
+    # Decompose the residualidual by SVD
+    output = _low_rank_decomposition(residual, reduced_rank=reduced_rank)
+    L, R, reduced_rank = output["L"], output["R"], output["reduced_rank"]
+    return R, L
+
+
+@torch.no_grad()
+def replace_lora_weights_loftq(
```

Comment:
Let's also document this method somewhere in the API section so users can easily refer to it. Maybe put it in a `## Utility` section with the [LoRA API](https://huggingface.co/docs/peft/package_reference/lora)?

Reply:
Thanks for the suggestion, I added an entry here: https://moon-ci-docs.huggingface.co/docs/peft/pr_1543/en/package_reference/lora#peft.replace_lora_weights_loftq